### ISSUE TITLE
Take control over the location of the search engine temp folders

### ISSUE THREAD

**Issue by bernt-matthias** [compomics/searchgui#272]
This seems to happen here: https://github.com/compomics/searchgui/blob/4ad38abe49c1a32be45543a2c9e00f0bab7d2d1f/src/main/java/eu/isas/searchgui/processbuilders/TandemProcessBuilder.java#L185

Consequence is that SearchGUI writes into the conda environment or the container which is a bad idea or strictly impossible. 

This leads to the workaround that the complete SearchGUI installation is currently copied before each run (when the conda wrapper script is called).

**Comment by bernt-matthias** [comment]
The SearchGUI log files are just another case. The following comes from a bioconda build: 

```
Please see the SearchGUI log file: /usr/local/share/searchgui-4.0.6-1/resources/SearchGUI.log
```

Certainly a path where SearchGUI should not write.

**Comment by bernt-matthias** [comment]
Actually on reasonable installations it should not be necessary to provide  `xTandem_directory`. The OS should know where to find it.

**Comment by CarlosHorro** [comment]
Hi @bernt-matthias ,

I'm afraid this is complicated to change. Many of the different search engines contained into SG need their input files or write their results in the same folder than they are installed, and sometimes that cannot be changed by us ie. using a custom temp folder (it is controlled by themselves).

It is not a new problem, and not an easy one to solve...

**Comment by bernt-matthias** [comment]
I see. Maybe we can improve this in little steps over time. Could you create a list of tools that have this requirement? Then we could check 1 by 1 if we can request a change upstream or if we can trick the software by symlinks.

**Comment by hbarsnes** [comment]
We can already control some of the search engine temp folders. For example from the temp settings (https://github.com/compomics/searchgui/wiki/SearchCLI#pathsettingscli) we should be able to do this for Comet and Andromeda.

Not sure how well-tested these options are though or how easy it would be to do the same for additional search engines. Pretty sure some of them do not allow us to change their internal temp folders. But I would need to dig into this in more detail to be sure.

> Then we could check 1 by 1 if we can request a change upstream or if we can trick the software by symlinks.

Keep in mind that not all of the search engines are still being maintained. So requesting upstream changes may not be that easy...

> The SearchGUI log files are just another case.

But shouldn't these already be redirected as mentioned here: https://github.com/compomics/searchgui/issues/211?

**Comment by hbarsnes** [comment]
Maybe I was too pessimistic with regards to which search engines we can take over the control of the temp folders. :)

I had a quick look and the only two with a firm 'no' seems to be OMSSA and MetaMorpheus. Both related to the modification parameters (and for MetaMorpheus also the enzymes). MetaMorpheus is still in development and we should hopefully be able to get them to extend their command line to support alternative locations of the internal files used for these particular settings. OMSSA is another matter, as the developers stopped maintaining it years ago.

In addition to these two, Andromeda, while having a temp folder we already control, it also has a local "conf" folder that I'm not sure if we can change.

As for all of the others, I think we should be able to control the temp folders used without too much trouble. But of course this all needs to be implemented and thoroughly tested before we can know for sure.

**Comment by bernt-matthias** [comment]
> I had a quick look and the only two with a firm 'no' seems to be OMSSA and MetaMorpheus.

Not sure about this: https://arxiv.org/pdf/q-bio/0406002.pdf: _The source code for OMSSA is in the public      domain      and      is      part      of      the      NCBI      C++      toolkit,      available      at      ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/CURRENT/_ .. which seems to be true. 

If we can compile it from source .. we can patch it :) 

For MetaMorpheus the release notes for the latest release state _Added a new optional command-line flag, --mmsettings, which can be used to specify a directory to read/write MetaMorpheus settings from/to. This enables the use of MetaMorpheus in readonly filesystems_

**Comment by hbarsnes** [comment]
> Not sure about this: https://arxiv.org/pdf/q-bio/0406002.pdf: The source code for OMSSA is in the public domain and is part of the NCBI C++ toolkit, available at ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/CURRENT/ .. which seems to be true.

Sure, but recompiling it apparently is a nightmare. I have never tried myself, but from what I understand it requires recompiling a lot more than just OMSSA.

> For MetaMorpheus the release notes for the latest release state Added a new optional command-line flag, --mmsettings, which can be used to specify a directory to read/write MetaMorpheus settings from/to. This enables the use of MetaMorpheus in readonly filesystems

Ah, did not know that! I guess they just forgot to update the command line wiki (https://github.com/smith-chem-wisc/MetaMorpheus/wiki/Using-the-Command-Line).

**Comment by hbarsnes** [comment]
A quick update on this issue. I've just released a new version of SearchGUI that allows the user to take control over the temp folders for X! Tandem, MS Amanda, MS-GF+, Comet, MetaMorpheus and Novor via the new search_engine_temp option which has been added to the PathSettingsCLI command line: https://github.com/compomics/searchgui/wiki/SearchCLI#pathsettingscli. (Or via the GUI: Edit > Resource Settings.)

The status of search engine temp folders is therefore as follows:
- X! Tandem, MS Amanda, MS-GF+, Comet, MetaMorpheus and Novor: temp folder can be set
- MyriMatch and DirecTag: does not use a temp folder
- Tide (or rather the tide-index): we can control the temp folder, but not the output folder for the index as we can only set the folder name (see https://crux.ms/commands/tide-index.html)
- Andromeda: we can control the temp folder, but not the conf folder where the input settings are stored
- OMSSA: does not use a temp folder, but relies on two local files for ptm details (mods.xml and usermods.xml), the latter we need to update to support user ptms

So for eight out of eleven search engines we should now be ok. :)

For Tide we can contact the crux developers (https://groups.google.com/g/crux-users) to see what can be done. They are very responsive and may be convinced to make the required changes. I will take care of this.

Andromeda is a different matter, as here the documentation is poor or missing and there seems to be little interest to support the standalone version of Andromeda (https://groups.google.com/g/andromeda-list).

Finally, there is OMSSA. I know there is the potential option to recompile it from the source code, but I'm not sure whether it would be worth it? Are there any other options we could try first?

**Comment by bernt-matthias** [comment]
> Finally, there is OMSSA. I know there is the potential option to recompile it from the source code, but I'm not sure whether it would be worth it?

I tried, but failed. Maybe one could ask ncbi support. Maybe they see the advantage of having this in bioconda even if they consider the project as deprecated.

> Are there any other options we could try first?

Not that I'm aware of.

**Comment by hbarsnes** [comment]
> Are there any other options we could try first?

Turns out there were! I had a closer look at the many OMSSA command line options and found one that we actually did not use:

```
 -mux <String>
   file containing user modification data
   Default = `usermods.xml'
```

And unlike for Tide, we can actually provide a path and not just a file/folder name.

The result is that we can now also control the OMSSA temp folder!

Which only leaves Tide and Andromeda. I still haven't gotten around to contacting the Tide/Crux developers yet. Will keep you posted.

**Comment by hbarsnes** [comment]
Tide can now also be added to the search engines where we can control the temp folders via the search_engine_temp option for PathSettingsCLI!

Turns out that the required Tide options were already there, it was just unclear from the descriptions that it was possible to provide paths and not just folder names. I just deployed SearchGUI v4.0.12 with this fix implemented.

So that only leaves Andromeda...
